tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslContext.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics;
+
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.security.KeyManagementException;
+import java.security.SecureRandom;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLContextSpi;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLServerSocketFactory;
+import javax.net.ssl.SSLSessionContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+
+final class InstrumentedSslContext extends SSLContext {
+
+    private final SSLContext context;
+    private final String name;
+
+    InstrumentedSslContext(SSLContext context, TaggedMetricRegistry metrics, String name) {
+        super(new InstrumentedSslContextSpi(context, metrics, name), context.getProvider(), context.getProtocol());
+        this.context = context;
+        this.name = name;
+    }
+
+    @Override
+    public String toString() {
+        return "InstrumentedSSLContext{delegate=" + context + ", name=" + name + '}';
+    }
+
+    private static final class InstrumentedSslContextSpi extends SSLContextSpi {
+
+        private final SSLContext context;
+        private final TaggedMetricRegistry metrics;
+        private final String name;
+
+        InstrumentedSslContextSpi(
+                // This must be the delegate context, passing an InstrumentedSSLContext
+                // will result in infinite recursion.
+                SSLContext context,
+                TaggedMetricRegistry metrics,
+                String name) {
+            this.context = context;
+            this.metrics = metrics;
+            this.name = name;
+        }
+
+        @Override
+        protected void engineInit(
+                KeyManager[] keyManagers,
+                TrustManager[] trustManagers,
+                SecureRandom secureRandom) throws KeyManagementException {
+            context.init(keyManagers, trustManagers, secureRandom);
+        }
+
+        @Override
+        protected SSLSocketFactory engineGetSocketFactory() {
+            return new InstrumentedSslSocketFactory(context.getSocketFactory(), metrics, name);
+        }
+
+        @Override
+        protected SSLServerSocketFactory engineGetServerSocketFactory() {
+            return new InstrumentedSslServerSocketFactory(context.getServerSocketFactory(), metrics, name);
+        }
+
+        @Override
+        protected SSLEngine engineCreateSSLEngine() {
+            return InstrumentedSslEngine.instrument(context.createSSLEngine(), metrics, name);
+        }
+
+        @Override
+        protected SSLEngine engineCreateSSLEngine(String host, int port) {
+            return InstrumentedSslEngine.instrument(context.createSSLEngine(host, port), metrics, name);
```

Comment:
we can probably infer this from the handshakes, but it might be interesting to have a counter of live `SSLEngine`s that have not been closed inbound/outbound.

Reply:
That would be interesting. This PR has grown a bit larger than I intended, I can investigate adding a counter in a follow-up.